File: lib/models/equipment_connection_model.dart
import 'package:flutter/material.dart';

class EquipmentConnectionModel {
  final int no;
  final DateTime regDate;
  final String? code;
  final String line;
  final String equipment;
  final String workType;
  final String dataType;
  final String connectionType;
  final String status;
  final String detail;
  final DateTime? startDate;
  final DateTime? completionDate;
  final String remarks;
  final bool isModified;
  final bool isNew;

  EquipmentConnectionModel({
    required this.no,
    required this.regDate,
    this.code,
    required this.line,
    required this.equipment,
    required this.workType,
    required this.dataType,
    required this.connectionType,
    required this.status,
    required this.detail,
    this.startDate,
    this.completionDate,
    this.remarks = '',
    this.isModified = false,
    this.isNew = false,
  });

  EquipmentConnectionModel copyWith({
    int? no,
    DateTime? regDate,
    String? code,
    String? line,
    String? equipment,
    String? workType,
    String? dataType,
    String? connectionType,
    String? status,
    String? detail,
    DateTime? startDate,
    DateTime? completionDate,
    String? remarks,
    bool? isModified,
    bool? isNew,
  }) {
    return EquipmentConnectionModel(
      no: no ?? this.no,
      regDate: regDate ?? this.regDate,
      code: code ?? this.code,
      line: line ?? this.line,
      equipment: equipment ?? this.equipment,
      workType: workType ?? this.workType,
      dataType: dataType ?? this.dataType,
      connectionType: connectionType ?? this.connectionType,
      status: status ?? this.status,
      detail: detail ?? this.detail,
      startDate: startDate ?? this.startDate,
      completionDate: completionDate ?? this.completionDate,
      remarks: remarks ?? this.remarks,
      isModified: isModified ?? this.isModified,
      isNew: isNew ?? this.isNew,
    );
  }

  factory EquipmentConnectionModel.fromJson(Map<String, dynamic> json) {
    return EquipmentConnectionModel(
      no: json['no'] ?? 0,
      regDate: json['regDate'] != null ? DateTime.parse(json['regDate']) : DateTime.now(),
      code: json['code'],
      line: json['line'] ?? '',
      equipment: json['equipment'] ?? '',
      workType: json['workType'] ?? '',
      dataType: json['dataType'] ?? '',
      connectionType: json['connectionType'] ?? '',
      status: json['status'] ?? '',
      detail: json['detail'] ?? '',
      startDate: json['startDate'] != null ? DateTime.parse(json['startDate']) : null,
      completionDate: json['completionDate'] != null ? DateTime.parse(json['completionDate']) : null,
      remarks: json['remarks'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'no': no,
      'regDate': regDate.toIso8601String(),
      'code': code,
      'line': line,
      'equipment': equipment,
      'workType': workType,
      'dataType': dataType,
      'connectionType': connectionType,
      'status': status,
      'detail': detail,
      'startDate': startDate?.toIso8601String(),
      'completionDate': completionDate?.toIso8601String(),
      'remarks': remarks,
    };
  }

  // 작업유형 목록
  static List<String> workTypes = [
    'MES자동실적',
    'MES자동투입',
    'SPC',
    '설비조건데이터',
  ];

  // 데이터유형 목록
  static List<String> dataTypes = [
    'CSV',
    'PLC',
  ];

  // 연동유형 목록
  static List<String> connectionTypes = [
    'DataAgent',
    'X-DAS',
    'X-SCADA',
  ];

  // 상태 목록
  static List<String> statusTypes = [
    '대기',
    '진행중',
    '완료',
    '중단',
  ];
} 

File: backend/models/equipmentConnectionModel.js
const mongoose = require('mongoose');

const equipmentConnectionSchema = new mongoose.Schema({
  no: { type: Number, required: true },
  code: { type: String, required: true, unique: true },
  regDate: { type: Date, required: true, default: Date.now },
  line: { type: String, required: true },
  equipment: { type: String, required: true },
  workType: { type: String, required: true }, // MES 자동투입, SPC, 설비조건데이터, 기타
  dataType: { type: String, required: true }, // PLC, CSV, 기타
  connectionType: { type: String, required: true }, // DataAgent, X-DAS, X-SCADA, 기타
  status: { type: String, required: true }, // 대기, 진행중, 완료, 보류
  detail: { type: String, default: '' },
  startDate: { type: Date },
  completionDate: { type: Date },
  remarks: { type: String, default: '' },
  isDeleted: { type: Boolean, default: false },
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now },
}, { 
  timestamps: true,
  collection: 'connection' // 컬렉션 이름을 connection으로 수정
});

// 인덱스 설정
equipmentConnectionSchema.index({ line: 1 });
equipmentConnectionSchema.index({ equipment: 1 });
equipmentConnectionSchema.index({ workType: 1 });
equipmentConnectionSchema.index({ dataType: 1 });
equipmentConnectionSchema.index({ connectionType: 1 });
equipmentConnectionSchema.index({ status: 1 });
equipmentConnectionSchema.index({ isDeleted: 1 });
equipmentConnectionSchema.index({ regDate: -1 });

// 검색을 위한 복합 인덱스
equipmentConnectionSchema.index({ 
  line: 'text', 
  equipment: 'text',
  workType: 'text',
  dataType: 'text',
  connectionType: 'text',
  detail: 'text',
  remarks: 'text'
});

// 새 문서 저장 전 처리
equipmentConnectionSchema.pre('save', function(next) {
  // 수정일시 자동 업데이트
  this.updatedAt = new Date();
  next();
});

const EquipmentConnectionModel = mongoose.model('EquipmentConnection', equipmentConnectionSchema);

module.exports = EquipmentConnectionModel; 